### PR TITLE
SONARJNKNS-294 Fix dead lock during Jenkins startup

### PR DESCRIPTION
**Pull request by henryju**
Seems that the combo GlobalConfiguration + Inititializer(PLUGINS_PREPARED) lead to a dead lock.
Looking at logstash-plugin, I changed it to JOB_LOADED

### CODE REVIEW

File: src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java [hunk 1]
```diff
@@ -72,25 +73,20 @@ public void setBuildWrapperEnabled(boolean enabled) {
...
-      }
-      // SonarPublisher might be null if Maven plugin is disabled or not installed
-      Jenkins j = Jenkins.getInstance();
-      DescriptorImpl publisher = j.getDescriptorByType(SonarPublisher.DescriptorImpl.class);
-      if (publisher != null && publisher.getDeprecatedInstallations() != null && publisher.getDeprecatedInstallations().length > 0) {
-
+    Optional<DescriptorImpl> publisherOpt = ExtensionList.lookup(SonarPublisher.DescriptorImpl.class).stream().findFirst();
+    // SonarPublisher might be missing if Maven plugin is disabled or not installed
+    publisherOpt.ifPresent(publisher -> {
+      if (publisher.getDeprecatedInstallations() != null && publisher.getDeprecatedInstallations().length > 0) {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: Code Smell Code Smell: Remove this use of "getDeprecatedInstallations"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=91&issues=AWKrOeFbUMN1uZ1OXWHY&open=AWKrOeFbUMN1uZ1OXWHY)

[ISSUE_KEY]: AWKrOeFbUMN1uZ1OXWHY

Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: Code Smell Code Smell: Remove this use of "getDeprecatedInstallations"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=91&issues=AWKrOeFbUMN1uZ1OXWHZ&open=AWKrOeFbUMN1uZ1OXWHZ)

[ISSUE_KEY]: AWKrOeFbUMN1uZ1OXWHZ

File: src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java [hunk 2]
```diff
@@ -72,25 +73,20 @@ public void setBuildWrapperEnabled(boolean enabled) {
...
+   * It will refuse to migrate if a SonarQube installation already exists in this class.
    * Migration will only be attempted once. 
    */
-  @Initializer(after = InitMilestone.PLUGINS_PREPARED)
-  public void migrate() {
-    if (migrated) {
+  @SuppressWarnings("deprecation")
+  @Initializer(after = InitMilestone.JOB_LOADED)
+  public void migrateData() {
+    if (dataMigrated) {
```
Comment (dbmeneses) [inline src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: The way this is done, it assumes that the method is not called concurrently. If that's not guaranteed, we could use an AtomicBoolean and do here `if (dataMigrated.getAndSet(true)) return`.

Comment (henryju) [reply src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: Looking at the logstash plugin, I think we can assume the method is called only once during server startup.

**Pull request merged**